### DISCUSSION TITLE
Favorite Java IDE

### DISCUSSION THREAD

**Discussion by salvatore-coppola** [eclipse-kura/kura#5198] [Polls]
Greetings to the Eclipse Kura community!

The Kura team is working on a complete overhaul of the tools and development environments involved in Kura development.

We would therefore like to hear from you what is your default/preferred IDE used in day-to-day development.

Thank you!

**Comment by mattdibi** [comment]
Others: anything **but** Eclipse IDE

**Comment by SteffenSenchyna** [comment]
@GregoryIvo, please release me from the eclipse prison.

**Reply by GregoryIvo** [reply]
You should share some screenshots/plugin list of your kura + IntelliJ setup 🤩

**Reply by SteffenSenchyna** [reply]
Using these two plugins I was able to have InteliJ recognize the manifest packages, the only downside is their can only be one source root for the project so the examples sub folder won't work properly.
![Screenshot 2024-04-23 at 7 36 33 PM](https://github.com/eclipse/kura/assets/105737839/fc7e8176-2a02-49a1-933e-8279057a8937)

**Reply by mattdibi** [reply]
> Using these two plugins I was able to have InteliJ recognize the manifest packages, the only downside is their can only be one source root for the project so the examples sub folder won't work properly. ![Screenshot 2024-04-23 at 7 36 33 PM](https://private-user-images.githubusercontent.com/105737839/325043271-fc7e8176-2a02-49a1-933e-8279057a8937.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM5Mzk5OTUsIm5iZiI6MTcxMzkzOTY5NSwicGF0aCI6Ii8xMDU3Mzc4MzkvMzI1MDQzMjcxLWZjN2U4MTc2LTJhMDItNDlhMS05MzNlLTgyNzkwNTdhODkzNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNFQwNjIxMzVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wY2Q4NjRkOTVhNjQ4MDgxMjgyNzU0OGU5NzRlZDI3Yzg2YTE4YmYxOWJhYzFjMDIwMThhNWY4ZmM0MGQ3NDVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ovl1Jw0oCQubseGzWQYGh_KAu5xsWFhGETTJX-NFmwo)

![image](https://github.com/eclipse/kura/assets/22748355/48e07c6f-ea32-46e3-8e4f-2b9a2e8f8629)

**Reply by mattdibi** [reply]
@SteffenSenchyna do you mind sharing your plugins configuration? I'm **really** interested in how you set this up....

**Reply by SteffenSenchyna** [reply]
Yeah for sure, so I actually got the full repo working in a single InteliJ project. Here are the steps I took, not sure if some of these steps are necessary as I am very new to Java/ OSGI
- Clone the repo and do the mvn clean install
- Opened the repo in Eclipse and set the target platform (Not sure if this is required)
- For the OSGI plugin I downloaded the Equinox SDK and had InteliJ point to it
![Screenshot 2024-04-24 at 1 27 58 AM](https://github.com/eclipse/kura/assets/105737839/d241a4df-9cec-4b68-9bf8-8af4550a2500)
![Screenshot 2024-04-24 at 1 28 46 AM](https://github.com/eclipse/kura/assets/105737839/8c90f676-200d-410d-8380-faeb9ca4baf6)
- The Eclipse PDE plugin I did not touch, and left as default settings, but with testing both of them are required for the IDE to recognize the full repo.
![Screenshot 2024-04-24 at 1 32 22 AM](https://github.com/eclipse/kura/assets/105737839/883c1250-4704-4647-9d3b-dd632c9c7899)
- Opened the repo in InteliJ with the two plugins installed, two options for importing the project should appear (Import Eclipse and Maven project), choose maven
- InteliJ will then index the root dir for dependencies but not recognize the Kura and Examples folder so these need to be imported as maven projects into the root project
![Screenshot 2024-04-24 at 1 36 33 AM](https://github.com/eclipse/kura/assets/105737839/431772cc-965a-4c8c-a48c-8f8982380cc5)
- This importing process will work the first time but when the project is closed and reopened the IDE will throw OSGI import error that the manifest imports are not in the classpath, this is just the IDE inspection throwing errors, following the steps in this [link](https://stackoverflow.com/questions/56674033/why-does-intellij-give-me-osgi-related-warnings-for-a-project-that-doesnt-use-a) you can disable the inspection cause the bundles still build with mvn clean install
- Then I just set up maven run configs for the individual modules
![Screenshot 2024-04-24 at 1 43 17 AM](https://github.com/eclipse/kura/assets/105737839/aad005b4-0125-4179-80e6-bc2f4bafe92e)
- I was able to modify the sensehat driver and build it with this process, I can also share my .idea project files as well if that helps

**Reply by mattdibi** [reply]
@SteffenSenchyna Thanks for the throughout explanation!

> I can also share my .idea project files as well if that helps

That would be awesome! I have [a bit of a history](https://github.com/eclipse/kura/pull/4573) of trying to move us away from Eclipse IDE now... you gave me new hope!

**Reply by SteffenSenchyna** [reply]
Here is the .idea project files for 
- Root dir: [root.idea.zip](https://github.com/eclipse/kura/files/15090757/root.idea.zip)
- Kura dir: [kura.idea.zip](https://github.com/eclipse/kura/files/15090773/kura.idea.zip)
I can also zip the whole project I have working right now but its 3.5 GB and its well over the attachment size limit. The OSGI plugin creates META-INF dirs on import that will be missing from the project.

**Reply by mattdibi** [reply]
> I can also zip the whole project I have working right now but its 3.5 GB and its well over the attachment size limit. The OSGI plugin creates META-INF dirs on import that will be missing from the project.

No worries. I think this is enough material for us to start looking into this. Thanks for sharing 😄

**Reply by SteffenSenchyna** [reply]
Sounds good, feel free to reach out if you have any questions.

**Reply by mattdibi** [reply]
Hello there!

I was able to perform a couple of tests and reproduce the setup **with two major differences**.

1. I didn't install a separate Equinox SDK instance but used the one shipped with Kura.

<img width="616" alt="IMG-24-04-2024-22-52-40" src="https://github.com/eclipse/kura/assets/22748355/4ef1e633-663b-4556-b256-c8ece8a0f602">
<img width="616" alt="IMG-24-04-2024-22-53-07" src="https://github.com/eclipse/kura/assets/22748355/c03bd724-6265-4704-a16a-afdb6cd58bf7">

2. I didn't use Eclipse for setting the target platform (in fact I didn't install Eclipse IDE _at all_ on my machine). I, instead, pointed the PDE partial plugin manually to the target platform like so:

<img width="776" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/e8df0ec6-26ee-421a-b7aa-786f36f2cc97">

This is a **critical** difference: it means that we don't need Eclipse **at all** for this setup to work! The dependency on the Eclipse IDE files of [my previous attempt](https://github.com/eclipse/kura/pull/4573) is what ultimately tanked it (it meant that either we added the Eclipse IDE project files to version control **or** users would have to install both Eclipse and VSCode to make it all work). This looks **really** promising!

Couple of more things:

- I was able to setup the formatter importing the setup from the `kura/setups` folder (Formatter and Cleanup can be merged in the same formatter profile):

![Pasted image 20240425143956](https://github.com/eclipse/kura/assets/22748355/6ddbc1d7-39b8-4fbf-bf06-c63068b1d830)

- I had to [follow the link](https://stackoverflow.com/questions/56674033/why-does-intellij-give-me-osgi-related-warnings-for-a-project-that-doesnt-use-a) you provided to disable the OSGi inspection thingy since it was throwing errors on the logger imports

I'm currently facing an issue with the unit tests. When I try to build and run a unit test from the IntelliJ UI it breaks with:

<img width="957" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/e4435325-85a3-4c97-94a9-7a2b9056c0f3">

<img width="957" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/f4a6306c-366e-441c-842a-79fee1087dfe">

<img width="961" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/0582800a-a148-4c88-a457-9c3249013b6a">

**Update**: after fixing the manifest path in the module facets all build fail with the following:

<img width="915" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/daa2878b-be58-4b8b-beaf-008aea5442dd">

**Update 2**: apparently reloading the maven project fixed it?!?!? (I was testing OpenJDK 8)

<img width="971" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/dc60633d-668a-4a34-a0ac-9a0f2018b86e">

I'm curious: @SteffenSenchyna can you run tests correctly with your setup?

In conclusion: a lot more testing is required but IntelliJ looks promising. To officially support IntelliJ as a Kura IDE I think we need _at the very minimum_:
- [X] Diagnostics support (already working)
- [X] Autocompletion support (already working)
- [X] Formatter support (already setup)
- [X] Unit test running and debugging (after Maven project reload... ?)
- [x] Remote debugging
- [ ] Stability testing

---

Additional test setup info:

<img width="579" alt="immagine" src="https://github.com/eclipse/kura/assets/22748355/5bd56ead-df96-4e96-b22a-c7cbe44b79eb">

It's just an Ubuntu 22.04LTS x86_64 VM I setup for this.

**Reply by SteffenSenchyna** [reply]
Hi @mattdibi that's very good to hear that you got the project working as well.  I also followed your steps with the target definitions and was able to load the project correctly as well. I was able to execute unit tests from the IDE as well
![unit_test](https://github.com/eclipse/kura/assets/105737839/1cdbfc73-a212-42b1-99b8-b16a5066a1b5)

However while I taking the screen shots intelij crashed and the config I had for setting up the project was not saved. After playing with the settings I found that right clicking on the test dir's pom and clicking resolve workspace resolved the missing manifest error.
![resolve_workspace](https://github.com/eclipse/kura/assets/105737839/e4464c67-a269-4eab-9ece-e44513067784)

Also @GregoryIvo and I did get remote OSGI bundle debugging running on a RPI 4, I unfortunately did not take screenshots of the config for that set up but I can confirm that remote OSGI debugging is possible. 

I have been working with trying to get the maven archetype project setup working and I have a set up working but it is using the esf-maven-archetype.jar not kura. InteliJ also doesn't support external archetype catalogue urls however once installed into the local maven repo I can create a new project directly from InteliJ.
![archetype](https://github.com/eclipse/kura/assets/105737839/0cb62f5a-0e56-4bb3-906d-bf0cc193d287)
![archetype_full_config](https://github.com/eclipse/kura/assets/105737839/84297914-5084-40a3-a9af-a1a207fe803b)

The issue with the archetype project is how the pom's are set up. InteliJ is set up to only a single project at a time this excerpt from this [article](https://www.jetbrains.com/help/idea/migrating-from-eclipse-to-intellij-idea.html#user_interface) of migrating from Eclipse to InteliJ.

> No workspace﻿
The first thing you'll notice when launching IntelliJ IDEA is that it has no workspace concept. This means that you can work with only one project at a time. While in Eclipse you normally have a set of projects that may depend on each other, in IntelliJ IDEA you have a single project that consists of a set of modules. If you have several unrelated projects, you can open them in separate windows. If you still want to have several unrelated projects opened in one window, as a workaround you can [configure them as modules](https://www.jetbrains.com/help/idea/creating-and-managing-modules.html#import-module).

So InteliJ does not interpret project structure correctly. The workaround I currently have is that once the full project is created I create a new project in the bundles dir and I can create new bundles importing existing kura packages. Having an InteliJ maven clean install run config in the bundles parent folder still allows for the creation of dp files in the features target folder. While this is using esf related resources, this info might be of use for creating a kura maven archetype.

Also there is another plugin of note that I found that could be useful however it don't seem to be actively maintained:
- [Eclipser](https://plugins.jetbrains.com/plugin/7153-eclipser) Converts Eclipse launch configurations into IntelliJ IDEA configurations

**Reply by mattdibi** [reply]
Hello there @SteffenSenchyna, a brief update: we officially started an internal evaluation of IntelliJ IDEA for Kura development. Currently involved in the process: me and @pierantoniomerlino. Our goal is to have something reproducible and stable enough that we can add it as one of our officially supported IDEs. Stay tuned! ;)

**Reply by SteffenSenchyna** [reply]
Hi @mattdibi, that is amazing news to hear. If you need help testing IDE setups please let me know I would be more then happy to help, as well as add to the documentation for new developer experience. Like I said previously I do not have any experience with Java or OSGI, so I can document any issues I encountered in getting set up to help out future devs who want to jump into Kura development.

**Reply by mattdibi** [reply]
Hello there @SteffenSenchyna, I apologise for the radio silence: we've been *really* busy with other projects emergencies and the new releases of Kura. Only last week I was able to resume the work on IntelliJ IDEA integration.

Unfortunately we've got bad news about it.

Both me and @pierantoniomerlino, after the initial experiments, were not able to set up the IDE in a way that allows us to run and debug unit tests. In [my previous message](https://github.com/eclipse/kura/discussions/5198#discussioncomment-9233987) I stated that it was working for me... but for whatever reason it was not reproducible nor stable (it stopped working almost immediately). Unfortunately the lack of unit test debugging/running is a showstopper for our adoption. I tried my best multiple times on multiple machines but I clearly lack the skillset to fix the issue and can't get out of this hole on my own.

I think there's something we're missing.

As a last resort/attempt I would like to ask you if you can describe *in even more detail* how do you setup IntelliJ idea for Kura development on a clean machine (or even a screen recording of the setup). Hopefully this would clean things up for us. Feel free to decline if you don't have the time obviously.

**Reply by SteffenSenchyna** [reply]
Hi @mattdibi no worries I will try and reproduce my steps from a clean setup and see if I can reproduce my previous success. Before I try this out I would just like to confirm which version of Java should be used 8 or 17?

**Reply by mattdibi** [reply]
Hi @SteffenSenchyna,

thanks 🙏  this is really appreciated. If you ever come to Italy we'll cover all your beer expenses 😄 

> which version of Java should be used 8 or 17?

17 since this is what we're moving to.

**Reply by SteffenSenchyna** [reply]
Hi @mattdibi just checking in to say I am still working on this, this time around InteliJ seems to be even less stable for loading the project dependencies properly, not sure what changed.

**Reply by mattdibi** [reply]
Hi @SteffenSenchyna, thanks for chiming in and taking the time to work on this. Timing is not an issue, don't worry.

**Comment by gnunzi** [comment]
Hallo, I am reposting here the picture of the heater example under vscode from a parallel discussion.

I realized that vscode probably works already for new kura modules out-of-the box. This should be easy if the module is packaged as "bundle" like the heater example. For other packagings it might be a bit trickier.

Is this of any value?
In case, we could extend the documentation to explain how to use vscode for new modules.

![image](https://github.com/user-attachments/assets/93e9eeb6-0dfb-4b11-98c4-09e5b3d99fd2)

**Reply by mattdibi** [reply]
Hello there @gnunzi,

can you share what VSCode plugins did you use for this setup? As I already shared in the other discussion, we already experimented with VSCode [here](https://github.com/eclipse-kura/kura/pull/4573). Ultimately we abandoned the idea since, despite working quite well, it still required a functioning Eclipse IDE installation to work (Eclipse PDE support in VSCode requires `.classpath` and `.project` files which are only generated by Eclipse IDE).

**Reply by gnunzi** [reply]
Hi @mattdibi,
I am using the "standard" plugins. Down below is a screenshot from a codespace instance.

The Eclipse PDE is presumably needed by the "eclipse-plugin" modules. The "bundle" modules are handle by the maven-bundle-plugin (bnd tools) and need an _explicit_ declaration of the dependecies in the pom (like the heater example) - no need for the Eclipse PDE.
This looks perfectly fine to me for the scenario of creating a _new_ module, since you basically just need the kura.api bundle and the osgi annotations+interfaces bundles (like in the heater example). The bnd tools are to my knowledge also the easiest way to develop in OSGi (in particular the automatic generation of the manifest and of the metadata from annotations are a booster).

The only trick for a "new module scenario" is to open the module's directory only. Then you have a clean IDE with hints to your module only. The parent pom is resolved from the local m2 repo, no need to have it open in your IDE (of course you need to run mvn install on kura's pom first).

The handling of the versions in the dependencies needs some thinking. One could use open-ended release versions (maybe ok for a "new module scenario"). Or one can use the snapshot versions, but then they should be updated at each release.

You can see an example in https://github.com/gnunzi/kura-fork/blob/example-new-module/kura/example-new-module/pom.xml.

You can see it is reported as "bundle" module. Then it gets compiled, packaged and installed.
![image](https://github.com/user-attachments/assets/c8f3e6e5-b0fb-47b6-b284-61018f2452ee)

I find this approach particularly useful for new modules, since there is almost no learning curve to start writing a new component. I actually created an article last year already on a similar procedure, but for docker deployment: https://github.com/gnunzi/kura-components-with-annotations/blob/main/README.md  (maybe outdated).

Here the plugins I used on a codespace for testing today:
![image](https://github.com/user-attachments/assets/cb7c0f8d-35a7-49ac-a4d5-2f3401eb1350)

**Reply by gnunzi** [reply]
Hi @mattdibi,

I realized that the same "trick" in vscode can be applied to exiting modules as well.
In this case one would better create a workspace file (not open directory), since it is possible there to declare mulitple modules to work on, like here below:

![image](https://github.com/user-attachments/assets/734cc682-d687-46f5-af09-c217323a9b66)

The packaging for exiting modules needs to be changed in this case to "jar" (only on those modules being currently worked on), because the exisitng manifest and osgi files are to be used (in "bundle" modules those files are auto-generated).
![image](https://github.com/user-attachments/assets/a0b0dec1-83c1-40e1-b479-27bc29a8face)

It works on my current setup. The existing manifest is in the jar together with the existing osgi files. It does not work for all modules yet, but it does work on single modules.

Of course, one could switch to "jar" for development and back to "eclipse-plugin" for pushing in order to keep the current pipeline working as before.

Below there is a screenshot. You can see the three modules recognized as Java and Maven projects. Code hints is working. Vscode fetches for the H2DbService class a binary file (lower part of the screen), presumably fetched from the local m2 repo. Of course, one could add the path to the the kura.api module in vscode, but this is not needed, since kura.api is only a dependency.
![image](https://github.com/user-attachments/assets/51db7545-4153-40a9-9da7-d1d073cdd6f4)

I can follow-up on this as soon as I cleaned-up my setting.

Nervetheless, for "new modules" the approach discussed previously (maven-bundle-plugin and autogenerated manifest) should be working already.

**Comment by mattdibi** [comment]
@SteffenSenchyna @gnunzi For those of you interested in the topic we just merged #6200 which migrates our archetype from Tycho to bnd (feedback welcome btw). [Link to the docs](https://eclipse-kura.github.io/kura/docs-develop/java-application-development/kura-addon-archetype/). 

Our goal is to start migrating Kura addons from Tycho to bnd until the whole project is fully migrated (hopefully in time for release 6.1.0).

Why is this relevant? Tycho was what tied us to Eclipse IDE, switching to bnd makes the build compatible with any IDE without any additional configuration/setup. We tested Visual Studio Code, IntelliJ IDEA and Eclipse IDE so far.

**Reply by SteffenSenchyna** [reply]
@mattdibi this is huge can't wait to test this out with IntelliJ IDEA.

**Reply by GregoryIvo** [reply]
I also can't wait to try it!

**Reply by SteffenSenchyna** [reply]
@mattdibi would the archetype work for developing cloud connections packages?

**Reply by mattdibi** [reply]
> @mattdibi would the archetype work for developing cloud connections packages?

Yes! @SteffenSenchyna

**Reply by SteffenSenchyna** [reply]
Tested it out with InteliJ and worked perfectly also I was able to import the base kura project as a module and it resolved everything as well. Stoked on this. I am assuming all the dps built with this won't be useable until the next major release of Kura/ESF?

**Reply by mattdibi** [reply]
> Tested it out with InteliJ and worked perfectly

Happy to hear that! 

> I am assuming all the dps built with this won't be useable until the next major release of Kura/ESF?

Correct. Please also keep in mind that we're deprecating the DPs for packaging Kura/ESF add-ons. We will continue accepting them for compatibility but all the new add-ons will use deb/rpm packages for installation. The archetype you tested should already display this behaviour (i.e. should build a deb package not a DP).
